### PR TITLE
Initial merge from dev to master

### PR DESCRIPTION
**Pull request by scottlovegrove**

### CODE REVIEW

File: .vscode/extensions.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (jankratochvilcz) [inline .vscode/extensions.json:1]: I wonder if this belongs in a shared git repo?

File: .vscodeignore [hunk 2]
```diff
@@ -0,0 +1,10 @@
+.vscode/**
```
Comment (jankratochvilcz) [inline .vscodeignore:1]: What's this file about? It seems to pretty much ignore all code in the solution :D

File: README.md [hunk 3]
```diff
@@ -1,2 +1,30 @@
...
+
+This extension contributes the following settings:
+
+* `todoist.accesstoken`: enable/disable this extension
+
+## Known Issues
+
+N/A
+
+## Release Notes
```
Comment (jankratochvilcz) [inline README.md:26]: There's a `CHANGELOG.md` file, would be worth to pick one way for the release notes and stick with it

File: package.json [hunk 4]
```diff
@@ -0,0 +1,100 @@
+{
+	"name": "todoist",
```
Comment (jankratochvilcz) [inline package.json:2]: We should call this e.g. `Todoist for VS Code` as that's the convention most of our addins follow

File: src/commands/authCommands.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
+import * as vscode from 'vscode';
+import { signInUrlAndState, exchangeToken } from './../todoistClient';
+import { getQueryStrings } from '../uriUtils';
+import * as userPreferencesService from './../services/userPreferencesService';
+import { CommandIds } from './commandIds';
+import { updateStatusBarItem } from '../services/statusBarService';
+import * as messagingService from '../services/messagingService';
+
+export /**
```
Comment (jankratochvilcz) [inline src/commands/authCommands.ts:9]: This way of  inserting the comment between the export statement and the function declaration is a bit unusual, is there a specific reason for it?

File: src/commands/authCommands.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import { updateStatusBarItem } from '../services/statusBarService';
+import * as messagingService from '../services/messagingService';
+
+export /**
+ * Generates the command for triggering the OAuth process to sign
+ * into Todoist
+ */
+    const signInCommand = () => vscode.commands.registerCommand(
+        CommandIds.signInCommandId,
+        async () => {
```
Comment (jankratochvilcz) [inline src/commands/authCommands.ts:15]: it might be worth separating the command definitions from the command execution bodies as that'd allow us to have a very clean file for executing all the commands and then simple functions for the command bodies. This approach could turn out to have some testability benefits :-)

File: src/commands/authCommands.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+export /**
+ * Generates the command for allowing the user to manually enter 
+ * the access token from todoist.com/prefs/integrations
+ */
+    const addTokenManuallyCommand = () => vscode.commands.registerCommand(
+        CommandIds.addTokenManuallyCommandId,
+        async () => {
+            const token = await vscode.window.showInputBox({
+                placeHolder: "Paste your token here"
```
Comment (jankratochvilcz) [inline src/commands/authCommands.ts:28]: We should aim to store the strings in a separate place. See Outlook add-in which has this solved

File: src/commands/authCommands.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        });
+
+export /**
+ * Generates the command for signing the user out of Todoist within VS Code.
+ * All this does is remove the access token from settings
+ */
+    const signOutCommand = () => vscode.commands.registerCommand(
+        CommandIds.signOutCommandId,
+        async () => {
+            await userPreferencesService.setAccessToken("");
```
Comment (jankratochvilcz) [inline src/commands/authCommands.ts:45]: Setting a null or undefined could be more expressive as we can use typescript's type system to explicitly differentiate the two states

File: src/commands/authCommands.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * The actual code that handles the auth URI.
+ * This needed to be in a separate method to allow for the 
+ * async calls to be made in an awaitable fashion
+ * 
+ * @param {vscode.Uri} uri
+ */
+const handleUriCallback = async (uri: vscode.Uri) => {
+    if (uri.path === "/auth-complete") {
```
Comment (jankratochvilcz) [inline src/commands/authCommands.ts:71]: Could be good to store all the URLs in one place as consts (I also think the Outlook add-in does this)

File: src/commands/commandIds.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+export class CommandIds {
+
+    //#region Task Commands
```
Comment (jankratochvilcz) [inline src/commands/commandIds.ts:3]: Is this IDE-agnostic? We should not force people into VS Code

File: src/commands/taskCommands.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
+import * as vscode from 'vscode';
```
Comment (jankratochvilcz) [inline src/commands/taskCommands.ts:1]: the feedback from authCommands applies here as well

File: src/commands/taskCommands.ts [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    const addTaskToProjectCommand = () => vscode.commands.registerCommand(
+        CommandIds.addTaskToProjectCommandId,
+        async () => {
+            if (!isSignedIn()) {
+                messagingService.showNotSignedInError();
+                updateStatusBarItem();
+                return;
+            }
+
+            const collectInputs = async () => {
```
Comment (jankratochvilcz) [inline src/commands/taskCommands.ts:58]: Generally it's be great to think about the overall solution structure, i.e. how can we reuse some of these functions in the future. It'll be very easy for files like this to grow and become very messy if we continue with this pattern

File: src/commands/taskCommands.ts [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+            const state = await collectInputs();
+
+            if (state.isFailure) {
+                messagingService.showError(`There was a problem with project selection`);
+                return;
+            }
+
+            if (state.currentProject.description) {
+                const taskAdded = await todoistClient.addTaskToProject(
+                    +state.currentProject.description,
```
Comment (jankratochvilcz) [inline src/commands/taskCommands.ts:97]: What's the `+` sign here?

File: src/extension.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+	context.subscriptions.push(addTaskToInboxCommand());
+	context.subscriptions.push(addTaskToProjectCommand());
+
+	const statusBarItem = createStatusBarItem();
+	statusBarItem.show();
+	context.subscriptions.push(statusBarItem);
+};
+
+
+export function deactivate() { }
```
Comment (jankratochvilcz) [inline src/extension.ts:26]: Is no cleanup needed?

File: src/helpers/contentHelpers.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
+import * as vscode from 'vscode';
```
Comment (jankratochvilcz) [inline src/helpers/contentHelpers.ts:1]: Can we solve this more systematically instead of introducing `helpers` folders and files?

File: src/helpers/contentHelpers.ts [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+export /**
+ * Checks whether the task content is valid. This is a simple
+ * check that just ascertains whether the content is empty or not
+ *
+ * @param {string} text
+ * @returns
+ */
+    const isValidContent = async (text: string) => {
+        return text !== "" && text !== undefined
```
Comment (jankratochvilcz) [inline src/helpers/contentHelpers.ts:11]: We're missing whitespace only validation here

File: src/helpers/multiStepInput.ts [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+    static async run(start: InputStep) {
+        const input = new MultiStepInput();
+        return input.stepThrough(start);
+    }
+
+    private current?: QuickInput;
+    private steps: InputStep[] = [];
+
+    private async stepThrough(start: InputStep) {
+        let step: InputStep | void = start;
```
Comment (jankratochvilcz) [inline src/helpers/multiStepInput.ts:52]: The code in this method is fairly hard to follow, can we somehow simplify it or make ti more declarative?

File: src/helpers/multiStepInput.ts [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+                    throw err;
+                }
+            }
+        }
+        if (this.current) {
+            this.current.dispose();
+        }
+    }
+
+    async showQuickPick<T extends QuickPickItem, P extends QuickPickParameters<T>>({ title, step, totalSteps, items, activeItem, placeholder, buttons, shouldResume }: P) {
```
Comment (jankratochvilcz) [inline src/helpers/multiStepInput.ts:79]: linebreaks 🗡️ 

Pls make sure your linter enforces line lengh, you should be getting error warnings in these cases :-)

File: src/helpers/multiStepInput.ts [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+        const disposables: Disposable[] = [];
+        try {
+            return await new Promise<T | (P extends { buttons: (infer I)[] } ? I : never)>((resolve, reject) => {
```
Comment (jankratochvilcz) [inline src/helpers/multiStepInput.ts:82]: This type is very hard to grasp

File: src/helpers/multiStepInput.ts [hunk 20]
```diff
@@ -0,0 +1,181 @@
+import {
```
Comment (jankratochvilcz) [inline src/helpers/multiStepInput.ts:1]: This file is generally a bit hard to understand without really digging into it, it seems to be due to the imperative UI building. Would be great to make the way the UI is built and how it transitions between states more imperative!

File: src/helpers/projectHelpers.ts [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+            });
+
+            return true;
+        }
+        catch{
+            return false;
+        }
+    };
+
+export const shouldResume = () => {
```
Comment (jankratochvilcz) [inline src/helpers/projectHelpers.ts:43]: 👀

File: src/services/authenticationService.ts [hunk 22]
```diff
@@ -0,0 +1,11 @@
+import * as userPreferencesService from './userPreferencesService';
+
+export /**
+ * Checks whether a user is signed in or not
+ *
+ * @returns [true] if the user is signed in
+ */
+    const isSignedIn = () => {
+        const accessToken = userPreferencesService.getAccessToken();
+        return !(accessToken === "" || accessToken === undefined);
```
Comment (jankratochvilcz) [inline src/services/authenticationService.ts:10]: The "" handling doesn't seem necessary, we could do with null or undefined. Could simplify some of the conditions

File: src/test/suite/extension.test.ts [hunk 23]
```diff
@@ -0,0 +1,15 @@
+import * as assert from 'assert';
+
+// You can import and use all API from the 'vscode' module
+// as well as import your extension to test it
+import * as vscode from 'vscode';
+// import * as myExtension from '../extension';
+
+suite('Extension Test Suite', () => {
```
Comment (jankratochvilcz) [inline src/test/suite/extension.test.ts:8]: 😅

File: .vscodeignore (hunk 2 above)
Comment (scottlovegrove) [reply .vscodeignore:1]: This file basically tells the package compiler what code should go into the extension. Given the output of a typescript project is actually javascript, it makes no sense packaging up the typescript files :)

File: package.json (hunk 4 above)
Comment (scottlovegrove) [reply package.json:2]: The name isn't what gets displayed anywhere, but I have updated the `displayName` property with this.

File: src/commands/authCommands.ts (hunk 5 above)
Comment (scottlovegrove) [reply src/commands/authCommands.ts:9]: I was using a Typescript doc comment extension in vs code and that's where it put the comments. Looks like that extension is rubbish 🙄

**Comment by scottlovegrove** [comment]
> While I'm unsure if react is applicable for this add-in, IMO Redux could clean up the side-effect handling

@jankratochvilcz React is 100% not applicable to this. What sort of side-effect handling are you talking about? I don't think Redux isn't really applicable to these extensions either.

File: src/commands/commandIds.ts (hunk 10 above)
Comment (scottlovegrove) [reply src/commands/commandIds.ts:3]: It probably is VS Code specific, but it's a comment! It's hardly _forcing_ people to use VS Code. If they use notepad, it's gonna make no difference to their coding experience at all. Just means if they _do_ use VS Code, the option to collapse that region is there.

File: src/extension.ts (hunk 14 above)
Comment (scottlovegrove) [reply src/extension.ts:26]: Nope. But you need to have that function for an extension for _if_ your extension does require cleanup (which ours doesn't).

File: src/services/authenticationService.ts (hunk 22 above)
Comment (scottlovegrove) [reply src/services/authenticationService.ts:10]: I've removed all these checks anyway, the `WorkspaceConfiguration` object has a `has()` method on it, which I can call to check whether a setting is set or not.

File: .eslintrc.json [hunk 24]
```diff
@@ -0,0 +1,19 @@
+{
```
[the remark this answers is not in the document]
Comment (scottlovegrove) [reply .eslintrc.json:1]: Have added this in and will apply any changes it requires.

File: src/helpers/multiStepInput.ts (hunk 20 above)
Comment (scottlovegrove) [reply src/helpers/multiStepInput.ts:1]: My fault for using a Microsoft generated helper file I guess. Just seemed easier to use their file that worked exactly as I wanted it than try and faff about and reinvent the wheel with my own code.

File: src/commands/authCommands.ts (hunk 7 above)
Comment (scottlovegrove) [reply src/commands/authCommands.ts:28]: This is not an easy task and is most definitely not comparable with the Outlook add-in. My recommendation for this is for it to be a post-release task, or at the very least, outside of this PR.

The outlook plugin is able to use a react library, this cannot be done, and whilst there is an [MS supplied solution](https://www.npmjs.com/package/vscode-nls), it requires much more configuration and setup than should be in this PR otherwise it will take ages to get merged.

File: src/commands/taskCommands.ts (hunk 13 above)
Comment (scottlovegrove) [reply src/commands/taskCommands.ts:97]: I've added a comment around this.

File: .vscode/extensions.json (hunk 1 above)
Comment (jankratochvilcz) [reply .vscode/extensions.json:1]: 100644

File: src/commands/authCommands.ts (hunk 7 above)
Comment (jankratochvilcz) [reply src/commands/authCommands.ts]: https://github.com/Doist/todoist-vs-code/pull/4/commits/de3cd2ba81ebb40b1650a33ff4d4627a48dff6eb

File: src/helpers/projectHelpers.ts (hunk 21 above)
Comment (jankratochvilcz) [reply src/helpers/projectHelpers.ts]: @scottlovegrove this code moved, but we still have it. Following the code, I'm still not totally sure why we need this, especially since we seem to be passing an empty function everywhere. What's the purpose of having these, could you help me shed some light on this?

File: src/helpers/multiStepInput.ts (hunk 20 above)
Comment (jankratochvilcz) [reply src/helpers/multiStepInput.ts]: Makes sense, I've added a comment and a blanket Issue to refactor at 20965c4

File: src/todoistClient.ts [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+import { getAccessToken } from './services/userPreferencesService'
+
+const apiBaseUrl = 'https://api.todoist.com/rest/v1/'
+const authoriseUrl = 'https://todoist.com/oauth/authorize'
+const tokenUrl = 'https://todoist.com/oauth/access_token'
+const projectsEndpoint = 'projects'
+const taskEndpoint = 'tasks'
+
+const clientID = 'f259fe20fbe24ac88650dcc08cbd5672'
+const clientSecret = '7d4a9394dddd48759c219bf937d6da3d'
```
Comment (scottlovegrove) [inline src/todoistClient.ts:12]: If it's not been raised as an issue already, we probably need to remove these 😄

**Comment by scottlovegrove** [comment]
All LGTM. I can't approve this PR cos I was the author :)

File: src/helpers/projectHelpers.ts (hunk 21 above)
Comment (scottlovegrove) [reply src/helpers/projectHelpers.ts]: Think of this as tombstoning from WP days. If you close the input wizard, and then open it up again, should you want it to resume from where you left off? Or start afresh. This is needed as part of the `multiStepInput` file from MS. We could refactor it all out.

File: src/todoistClient.ts (hunk 25 above)
Comment (jankratochvilcz) [reply src/todoistClient.ts]: Have added env files
